### ISSUE TITLE
Collect the list of missing commands

### ISSUE THREAD

**Issue by mijicd** [zio/zio-redis#68]
Redis API has been expanded since the library was designed. The purpose of this issue is to come with the list of missing commands, and decide which ones we are going to support.

**Comment by regis-leray** [comment]
### ACL
- [ ] ACL LOAD Reload the ACLs from the configured ACL file
- [ ] ACL SAVE Save the current ACL rules in the configured ACL file
- [ ] ACL LIST List the current ACL rules in ACL config file format
- [ ] ACL USERS List the username of all the configured ACL rules
- [ ] ACL GETUSER username Get the rules for a specific ACL user
- [ ] ACL SETUSER username [rule [rule ...]] Modify or create the rules for a specific  ACL user
- [ ] ACL DELUSER username [username ...] Remove the specified ACL users and the associated rules
- [ ] ACL CAT [categoryname] List the ACL categories or the commands inside a category
- [ ] ACL GENPASS [bits] Generate a pseudorandom secure password to use for ACL users
- [ ] ACL WHOAMI Return the name of the user associated to the current connection
- [ ] ACL LOG [count or RESET] List latest events denied because of ACLs in place
- [ ] ACL HELP Show helpful text about the different subcommands

### CLIENT
- [ ] CLIENT CACHING YES|NO Instruct the server about tracking or not keys in the next request
- [ ] CLIENT ID Returns the client ID for the current connection
- [ ] CLIENT KILL [ip:port] [ID client-id] [TYPE normal|master|slave|pubsub] [USER username] [ADDR ip:port] [SKIPME yes/no] Kill the connection of a client
- [ ] CLIENT LIST [TYPE normal|master|replica|pubsub] Get the list of client connections
- [ ] CLIENT GETNAME Get the current connection name
- [ ] CLIENT GETREDIR Get tracking notifications redirection client ID if any
- [ ] CLIENT PAUSE timeout Stop processing commands from clients for some time
- [ ] CLIENT REPLY ON|OFF|SKIP Instruct the server whether to reply to commands
- [ ] CLIENT SETNAME connection-name Set the current connection name
- [ ] CLIENT TRACKING ON|OFF [REDIRECT client-id] [PREFIX prefix [PREFIX prefix ...]] [BCAST] [OPTIN] [OPTOUT] [NOLOOP] Enable or disable server assisted client side caching support
- [ ] CLIENT UNBLOCK client-id [TIMEOUT|ERROR] Unblock a client blocked in a blocking command from a different connecti- [ ]n

### CLUSTER    
- [ ] CLUSTER ADDSLOTS slot [slot ...] Assign new hash slots to receiving node
- [ ] CLUSTER BUMPEPOCH Advance the cluster config epoch
- [ ] CLUSTER COUNT-FAILURE-REPORTS node-id Return the number of failure reports active for a given node
- [ ] CLUSTER COUNTKEYSINSLOT slot Return the number of local keys in the specified hash slot
- [ ] CLUSTER DELSLOTS slot [slot ...] Set hash slots as unbound in receiving node
- [ ] CLUSTER FAILOVER [FORCE|TAKEOVER] Forces a replica to perform a manual failover of its master.
- [ ] CLUSTER FLUSHSLOTS Delete a node's own slots information
- [ ] CLUSTER FORGET node-id Remove a node from the nodes table
- [ ] CLUSTER GETKEYSINSLOT slot count Return local key names in the specified hash slot
- [ ] CLUSTER INFO Provides info about Redis Cluster node state
- [ ] CLUSTER KEYSLOT key Returns the hash slot of the specified ke- [ ]
- [ ] CLUSTER MEET ip port Force a node cluster to handshake with another node
- [ ] CLUSTER MYID Return the node id
- [ ] CLUSTER NODES Get Cluster config for the node
- [ ] CLUSTER REPLICATE node-id Reconfigure a node as a replica of the specified master node
- [ ] CLUSTER RESET [HARD|SOFT] Reset a Redis Cluster node
- [ ] CLUSTER SAVECONFIG Forces the node to save cluster state on disk
- [ ] CLUSTER SET-CONFIG-EPOCH config-epoch Set the configuration epoch in a new node
- [ ] CLUSTER SETSLOT slot IMPORTING|MIGRATING|STABLE|NODE [node-id] Bind a hash slot to a specific node
- [ ] CLUSTER SLAVES node-id List replica nodes of the specified master node
- [ ] CLUSTER REPLICAS node-id List replica nodes of the specified master node
- [ ] CLUSTER SLOTS Get array of Cluster slot to node mappings- [ ]

### KEYS
- [x] DEL key [key ...] Delete a key
- [x] DUMP key Return a serialized version of the value stored at the specified key.
- [x] EXISTS key [key ...] Determine if a key exists
- [x] EXPIRE key seconds Set a key's time to live in seconds
- [x] EXPIREAT key timestamp Set the expiration for a key as a UNIX timestamp
- [x] KEYS pattern Find all keys matching the given pattern
- [x] MIGRATE host port key|"" destination-db timeout [COPY] [REPLACE] [AUTH password] [KEYS key [key ...]] Atomically transfer a key from a Redis instance to another one.
- [x] MOVE key db Move a key to another database
- [x] PERSIST key Remove the expiration from a key
- [x] PEXPIRE key milliseconds Set a key's time to live in milliseconds
- [x] PEXPIREAT key milliseconds-timestamp Set the expiration for a key as a UNIX timestamp specified in milliseconds
- [x] PTTL key Get the time to live for a key in milliseconds
- [x] RANDOMKEY Return a random key from the keyspace
- [x] RENAME key newkey Rename a key
- [x] RENAMENX key newkey Rename a key, only if the new key does not exist
- [x] RESTORE key ttl serialized-value [REPLACE] [ABSTTL] [IDLETIME seconds] [FREQ frequency] Create a key using the provided serialized value, previously obtained using DUMP.
- [x] TOUCH key [key ...] Alters the last access time of a key(s). Returns the number of existing keys specified.
- [x] TTL key Get the time to live for a key
- [x] TYPE key Determine the type stored at key
- [x] UNLINK key [key ...] Delete a key asynchronously in another thread. Otherwise it is just as DEL, but non blocking.
- [x] WAIT numreplicas timeout Wait for the synchronous replication of all the write commands sent in the context of the current connection
- [ ] SCAN cursor [MATCH pattern] [COUNT count] [TYPE type] Incrementally iterate the keys space
- [ ] FLUSHALL [ASYNC] Remove all keys from all databases
- [ ] FLUSHDB [ASYNC] Remove all keys from the current database


### GEO    
- [x] GEOADD key longitude latitude member [longitude latitude member ...] Add one or more geospatial items in the geospatial index represented using a sorted set
- [x] GEOHASH key member [member ...] Returns members of a geospatial index as standard geohash strings
- [x] GEOPOS key member [member ...] Returns longitude and latitude of members of a geospatial index
- [x] GEODIST key member1 member2 [m|km|ft|mi] Returns the distance between two members of a geospatial index
- [x] GEORADIUS key longitude latitude radius m|km|ft|mi [WITHCOORD] [WITHDIST] [WITHHASH] [COUNT count] [ASC|DESC] [STORE key] [STOREDIST key] Query a sorted set representing a geospatial index to fetch members matching a given maximum distance from a point
- [x] GEORADIUSBYMEMBER key member radius m|km|ft|mi [WITHCOORD] [WITHDIST] [WITHHASH] [COUNT count] [ASC|DESC] [STORE key] [STOREDIST key] Query a sorted set representing a geospatial index to fetch members matching a given maximum distance from a member



### HASHES
- [x] HDEL key field [field ...] Delete one or more hash fields
- [ ] HELLO protover [AUTH username password] [SETNAME clientname] switch Redis protocol
- [x] HEXISTS key field Determine if a hash field exists
- [x] HGET key field Get the value of a hash field
- [x] HGETALL key Get all the fields and values in a hash
- [x] HINCRBY key field increment Increment the integer value of a hash field by the given number
- [ ] HINCRBYFLOAT key field increment Increment the float value of a hash field by the given amount
- [x] HKEYS key Get all the fields in a hash
- [x] HLEN key Get the number of fields in a hash
- [x] HMGET key field [field ...] Get the values of all the given hash fields
- [ ] HMSET key field value [field value ...] Set multiple hash fields to multiple values
- [x] HSET key field value [field value ...] Set the string value of a hash field
- [x] HSETNX key field value Set the value of a hash field, only if the field does not exist
- [x] HSTRLEN key field Get the length of the value of a hash field
- [x] HVALS key Get all the values in a hash
- [x] HSCAN key cursor [MATCH pattern] [COUNT count] Incrementally iterate hash fields and associated values


### LIST
- [x] LINDEX key index Get an element from a list by its index
- [x] LINSERT key BEFORE|AFTER pivot element Insert an element before or after another element in a list
- [x] LLEN key Get the length of a list
- [x] LPOP key Remove and get the first element in a list
- [ ] LPOS key element [RANK rank] [COUNT num-matches] [MAXLEN len] Return the index of matching elements on a list
- [x] LPUSH key element [element ...] Prepend one or multiple elements to a list
- [x] LPUSHX key element [element ...] Prepend an element to a list, only if the list exists
- [x] LRANGE key start stop Get a range of elements from a list
- [x] LREM key count element Remove elements from a list
- [x] LSET key index element Set the value of an element in a list by its index
- [x] LTRIM key start stop Trim a list to the specified range
- [x] RPOP key Remove and get the last element in a list
- [x] RPOPLPUSH source destination Remove the last element in a list, prepend it to another list and return it
- [x] RPUSH key element [element ...] Append one or multiple elements to a list
- [x] RPUSHX key element [element ...] Append an element to a list, only if the list exists
- [x] BLPOP key [key ...] timeout Remove and get the first element in a list, or block until one is available
- [x] BRPOP key [key ...] timeout Remove and get the last element in a list, or block until one is available
- [ ] BRPOPLPUSH source destination timeout Pop an element from a list, push it to another list and return it; or block until one is available

### MEMORY
- [ ] MEMORY DOCTOR Outputs memory problems report
- [ ] MEMORY HELP Show helpful text about the different subcommands
- [ ] MEMORY MALLOC-STATS Show allocator internal stats
- [ ] MEMORY PURGE Ask the allocator to release memory
- [ ] MEMORY STATS Show memory usage details
- [ ] MEMORY USAGE key [SAMPLES count] Estimate the memory usage of a key


### HYPERLOGLOG
- [x] PFADD key element [element ...] Adds the specified elements to the specified HyperLogLog.
- [x] PFCOUNT key [key ...] Return the approximated cardinality of the set(s) observed by the HyperLogLog at key(s).
- [x] PFMERGE destkey sourcekey [sourcekey ...] Merge N different HyperLogLogs into a single one.

### SETS
- [x] SADD key member [member ...] Add one or more members to a set
- [x] SCARD key Get the number of members in a set
- [x] SDIFF key [key ...] Subtract multiple sets
- [x] SDIFFSTORE destination key [key ...] Subtract multiple sets and store the resulting set in a key
- [x] SINTER key [key ...] Intersect multiple sets
- [x] SISMEMBER key member Determine if a given value is a member of a set
- [x] SMEMBERS key Get all the members in a set
- [x] SMOVE source destination member Move a member from one set to another
- [x] SPOP key [count] Remove and return one or multiple random members from a set
- [x] SRANDMEMBER key [count] Get one or multiple random members from a set
- [x] SREM key member [member ...] Remove one or more members from a set
- [x] SUNION key [key ...] Add multiple sets
- [x] SUNIONSTORE destination key [key ...] Add multiple sets and store the resulting set in a key
- [x] SSCAN key cursor [MATCH pattern] [COUNT count] Incrementally iterate Set elements


### SORTED-SETS
- [x] BZPOPMIN key [key ...] timeout Remove and return the member with the lowest score from one or more sorted sets, or block until one is available
- [x] BZPOPMAX key [key ...] timeout Remove and return the member with the highest score from one or more sorted sets, or block until one is available
- [x] ZADD key [NX|XX] [CH] [INCR] score member [score member ...] Add one or more members to a sorted set, or update its score if it already exists
- [x] ZCARD key Get the number of members in a sorted set
- [x] ZCOUNT key min max Count the members in a sorted set with scores within the given values
- [x] ZINCRBY key increment member Increment the score of a member in a sorted set
- [x] ZINTERSTORE destination numkeys key [key ...] [WEIGHTS weight [weight ...]] [AGGREGATE SUM|MIN|MAX] Intersect multiple sorted sets and store the resulting sorted set in a new key
- [x] ZLEXCOUNT key min max Count the number of members in a sorted set between a given lexicographical range
- [x] ZPOPMAX key [count] Remove and return members with the highest scores in a sorted set
- [x] ZPOPMIN key [count] Remove and return members with the lowest scores in a sorted set
- [x] ZRANGE key start stop [WITHSCORES] Return a range of members in a sorted set, by index
- [x] ZRANGEBYLEX key min max [LIMIT offset count] Return a range of members in a sorted set, by lexicographical range
- [x] ZREVRANGEBYLEX key max min [LIMIT offset count] Return a range of members in a sorted set, by lexicographical range, ordered from higher to lower strings.
- [x] ZRANGEBYSCORE key min max [WITHSCORES] [LIMIT offset count] Return a range of members in a sorted set, by score
- [x] ZRANK key member Determine the index of a member in a sorted set
- [x] ZREM key member [member ...] Remove one or more members from a sorted set
- [x] ZREMRANGEBYLEX key min max Remove all members in a sorted set between the given lexicographical range
- [x] ZREMRANGEBYRANK key start stop Remove all members in a sorted set within the given indexes
- [x] ZREMRANGEBYSCORE key min max Remove all members in a sorted set within the given scores
- [x] ZREVRANGE key start stop [WITHSCORES] Return a range of members in a sorted set, by index, with scores ordered from high to low
- [x] ZREVRANGEBYSCORE key max min [WITHSCORES] [LIMIT offset count] Return a range of members in a sorted set, by score, with scores ordered from high to low
- [x] ZREVRANK key member Determine the index of a member in a sorted set, with scores ordered from high to low
- [x] ZSCORE key member Get the score associated with the given member in a sorted set
- [x] ZUNIONSTORE destination numkeys key [key ...] [WEIGHTS weight [weight ...]] [AGGREGATE SUM|MIN|MAX] Add multiple sorted sets and store the resulting sorted set in a new key
- [x] ZSCAN key cursor [MATCH pattern] [COUNT count] Incrementally iterate sorted sets elements and associated scores

### STRINGS
- [x] APPEND key value Append a value to a key
- [x] BITCOUNT key [start end] Count set bits in a string
- [x] BITFIELD key [GET type offset] [SET type offset value] [INCRBY type offset increment] [OVERFLOW WRAP|SAT|FAIL] Perform arbitrary bitfield integer operations on strings
- [x] BITOP operation destkey key [key ...] Perform bitwise operations between strings
- [x] BITPOS key bit [start] [end] Find first bit set or clear in a string
- [x] DECR key Decrement the integer value of a key by one
- [x] DECRBY key decrement Decrement the integer value of a key by the given number
- [x] GET key Get the value of a key
- [x] GETRANGE key start end Get a substring of the string stored at a key
- [x] GETBIT key offset Returns the bit value at offset in the string value stored at key
- [x] GETSET key value Set the string value of a key and return its old value
- [x] INCR key Increment the integer value of a key by one
- [x] INCRBY key increment Increment the integer value of a key by the given amount
- [x] INCRBYFLOAT key increment Increment the float value of a key by the given amount
- [x] MGET key [key ...] Get the values of all the given keys
- [x] MSET key value [key value ...] Set multiple keys to multiple values
- [x] MSETNX key value [key value ...] Set multiple keys to multiple values, only if none of the keys exist
- [x] PSETEX key milliseconds value Set the value and expiration in milliseconds of a key
- [x] SET key value [EX seconds|PX milliseconds|KEEPTTL] [NX|XX] Set the string value of a key
- [x] SETBIT key offset value Sets or clears the bit at offset in the string value stored at key
- [x] SETEX key seconds value Set the value and expiration of a key
- [x] SETNX key value Set the value of a key, only if the key does not exist
- [x] SETRANGE key offset value Overwrite part of a string at key starting at the specified offset
- [x] STRLEN key Get the length of the value stored in a key

### TRANSACTIONS
- [x] DISCARD Discard all commands issued after MULTI
- [x] EXEC Execute all commands issued after MULTI
- [x] MULTI Mark the start of a transaction block
- [x] UNWATCH Forget about all watched keys
- [x] WATCH key [key ...] Watch the given keys to determine execution of the MULTI/EXEC block

### LATENCY
- [ ] LATENCY DOCTOR Return a human readable latency analysis report.
- [ ] LATENCY GRAPH event Return a latency graph for the event.
- [ ] LATENCY HISTORY event Return timestamp-latency samples for the event.
- [ ] LATENCY LATEST Return the latest latency samples for all events.
- [ ] LATENCY RESET [event [event ...]] Reset latency data for one or more events.
- [ ] LATENCY HELP Show helpful text about the different subcommands.


----


- [ ] COMMAND Get array of Redis command details
- [ ] COMMAND COUNT Get total number of Redis commands
- [ ] COMMAND GETKEYS Extract keys given a full Redis command
- [ ] COMMAND INFO command-name [command-name ...] Get array of specific Redis command details
- [ ] CONFIG GET parameter Get the value of a configuration parameter
- [ ] CONFIG REWRITE Rewrite the configuration file with the in memory configuration
- [ ] CONFIG SET parameter value Set a configuration parameter to the given value
- [ ] CONFIG RESETSTAT Reset the stats returned by INFO
- [ ] DBSIZE Return the number of keys in the selected database
- [ ] DEBUG OBJECT key Get debugging information about a key
- [ ] DEBUG SEGFAULT Make the server crash

- [ ] INFO [section] Get information and statistics about the server
- [ ] LOLWUT [VERSION version] Display some computer art and the Redis version

- [ ] LASTSAVE Get the UNIX time stamp of the last successful save to disk


- [ ] AUTH password Authenticate to the server
- [ ] BGREWRITEAOF Asynchronously rewrite the append-only file
- [ ] BGSAVE [SCHEDULE] Asynchronously save the dataset to disk


   

- [ ] ECHO message Echo the given string

- [ ] EVAL script numkeys key [key ...] arg [arg ...] Execute a Lua script server side
- [ ] EVALSHA sha1 numkeys key [key ...] arg [arg ...] Execute a Lua script server side


- [ ] MODULE LIST List all modules loaded by the server
- [ ] MODULE LOAD path [ arg [arg ...]] Load a module
- [ ] MODULE UNLOAD name Unload a module
- [ ] MONITOR Listen for all requests received by the server in real time


- [ ] OBJECT subcommand [arguments [arguments ...]] Inspect the internals of Redis objects

- [ ] PING [message] Ping the server

- [ ] PSUBSCRIBE pattern [pattern ...] Listen for messages published to channels matching the given patterns
- [ ] PUBSUB subcommand [argument [argument ...]] Inspect the state of the Pub/Sub subsystem

- [ ] PUBLISH channel message Post a message to a channel
- [ ] PUNSUBSCRIBE [pattern [pattern ...]] Stop listening for messages posted to channels matching the given patterns
- [ ] QUIT Close the connection


- [ ] READONLY Enables read queries for a connection to a cluster replica node
- [ ] READWRITE Disables read queries for a connection to a cluster replica node
- [ ] ROLE Return the role of the instance in the context of replication



- [ ] SAVE Synchronously save the dataset to disk
- [ ] SCRIPT DEBUG YES|SYNC|NO Set the debug mode for executed scripts.
- [ ] SCRIPT EXISTS sha1 [sha1 ...] Check existence of scripts in the script cache.
- [ ] SCRIPT FLUSH Remove all the scripts from the script cache.
- [ ] SCRIPT KILL Kill the script currently in execution.
- [ ] SCRIPT LOAD script Load the specified Lua script into the script cache.
- [ ] SELECT index Change the selected database for the current connection
- [ ] SHUTDOWN [NOSAVE|SAVE] Synchronously save the dataset to disk and then shut down the server
- [ ] SINTERSTORE destination key [key ...] Intersect multiple sets and store the resulting set in a key
- [ ] SLAVEOF host port Make the server a replica of another instance, or promote it as master. Deprecated starting with Redis 5. Use REPLICAOF instead.
- [ ] REPLICAOF host port Make the server a replica of another instance, or promote it as master.
- [ ] SLOWLOG subcommand [argument] Manages the Redis slow queries log
- [ ] SORT key [BY pattern] [LIMIT offset count] [GET pattern [GET pattern ...]] [ASC|DESC] [ALPHA] [STORE destination] Sort the elements in a list, set or sorted set
- [ ] STRALGO LCS algo-specific-argument [algo-specific-argument ...] Run algorithms (currently LCS) against strings

- [ ] SUBSCRIBE channel [channel ...] Listen for messages published to the given channels
- [ ] SWAPDB index1 index2 Swaps two Redis databases
- [ ] SYNC Internal command used for replication
- [ ] PSYNC replicationid offset Internal command used for replication
- [ ] TIME Return the current server time
- [ ] UNSUBSCRIBE [channel [channel ...]] Stop listening for messages posted to the given channels


- [ ] XINFO [CONSUMERS key groupname] [GROUPS key] [STREAM key] [HELP] Get information on streams and consumer groups
- [ ] XADD key ID field value [field value ...] Appends a new entry to a stream
- [ ] XTRIM key MAXLEN [~] count Trims the stream to (approximately if '~' is passed) a certain size
- [ ] XDEL key ID [ID ...] Removes the specified entries from the stream. Returns the number of items actually deleted, that may be different from the number of IDs passed in case certain IDs do not exist.
- [ ] XRANGE key start end [COUNT count] Return a range of elements in a stream, with IDs matching the specified IDs interval
- [ ] XREVRANGE key end start [COUNT count] Return a range of elements in a stream, with IDs matching the specified IDs interval, in reverse order (from greater to smaller IDs) compared to XRANGE
- [ ] XLEN key Return the number of entires in a stream
- [ ] XREAD [COUNT count] [BLOCK milliseconds] STREAMS key [key ...] id [id ...] Return never seen elements in multiple streams, with IDs greater than the ones reported by the caller for each stream. Can block.
- [ ] XGROUP [CREATE key groupname id-or-$] [SETID key groupname id-or-$] [DESTROY key groupname] [DELCONSUMER key groupname consumername] Create, destroy, and manage consumer groups.
- [ ] XREADGROUP GROUP group consumer [COUNT count] [BLOCK milliseconds] [NOACK] STREAMS key [key ...] ID [ID ...] Return new entries from a stream using a consumer group, or access the history of the pending entries for a given consumer. Can block.
- [ ] XACK key group ID [ID ...] Marks a pending message as correctly processed, effectively removing it from the pending entries list of the consumer group. Return value of the command is the number of messages successfully acknowledged, that is, the IDs we were actually able to resolve in the PEL.
- [ ] XCLAIM key group consumer min-idle-time ID [ID ...] [IDLE ms] [TIME ms-unix-time] [RETRYCOUNT count] [FORCE] [JUSTID] Changes (or acquires) ownership of a message in a consumer group, as if the message was delivered to the specified consumer.
- [ ] XPENDING key group [start end count] [consumer] Return information and entries from a stream consumer group pending entries list, that are messages fetched but never acknowledged.

**Comment by regis-leray** [comment]
@mijicd  i used https://redis.io/commands to list all commands and only checked what is implemented

**Comment by mijicd** [comment]
@regis-leray amazing job, thank you! Lots of these were added in the meantime with redis 6, but it would be great if we can support them.

I planned to ship one small usability improvement and cut the 0.0.1 release but my work schedule postponed that. However, I think I'll have that by tomorrow and then we can crunch the rest of the stuff.

Goes without saying, I'd love to have you on board :). In fact, I think this ticket is pretty independent so you (and other volunteers) can start picking most of the commands. I can see how streaming, transactions and cluster are special (they should have separate tickets), but other commands are fairly easy to implement. WDYT?

**Comment by regis-leray** [comment]
@mijicd of course I will try my best to contribute.

so the plan is to support all missing commands?  
 - [ ] cluster, 
 - [ ] connection
 - [ ] pub/sub
 - [ ] scripting
 - [ ] server
 - [ ] streams

And i'm guessing we will add few missing commands for existing group.

is there any particular priority/order? or doesn't matter?

**Comment by mijicd** [comment]
I think @anovakovic01 is already working on streaming part, the rest of those are open for tackling. I'd prioritize adding the missing commands for existing groups though, just to make some "order" in releases :).

**Comment by mijicd** [comment]
I have closed 3 related tickets in favor of this one. From now on, please raise your hand for the command (or group of commands) you are interested in. @anovakovic01 already called "dibs" on streaming :).

**Comment by regis-leray** [comment]
i can try to implement `SCRIPTING`

**Comment by mijicd** [comment]
I have split this in multiple subtickets to make it a bit more hackathon friendly :).
